keps/sig-cli/4764-kubectl-watchlist/README.md

```diff
@@ -0,0 +1,757 @@
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Kubectl supports WatchList to list resources. This will allow requests from kubectl to hit the WatchCache,
+giving a performance boost and reducing impact on the kube-apiserver.
+
+## Motivation
+
+Currently `client-go` could enable `WatchList` by env and it works great, but kubectl is still a pain. 
+`kubectl` doest not have a cache mechanism like `client-go`, so it will do `list`  whenever it is used. Furthermore,
+request with pager params can not hit Cache of `kube-apiserver`, every request will go through into `etcd`,
+which could push up the load of `kube-apiserver` and make system unstable.
+
+In practice, our users have been writing script to unrestrainedly call `kubectl`, which could lead serious performance
+issue. As cluster admininstor, we have few method to eliminate it. APF maybe is an option, but it's hard to configure
+and don't solve the problem from the root. In the other side, if we are standing in the shoes of user, most of those 
+scripts are simple Bash or Python which just simply glue a few `kubectl` commands. It is too heavy and "not worth" 
+for user to refactor just "a few commands" into "a batch of complex code". So we are seeking an optimization for `kubectl`. With this optimization, all users of kubectl can get free performance improvement.
+
+
+
+### Goals
+
+* Introduce WatchList to `kubectl` without breaking any user interface.
+* Make `kubectl` hit the cache, thus reducing impact on `kube-apiserver`.
+
+### Non-Goals
+
+* Change kubectl functionality.
+
+## Proposal
+
+`kubectl` commands like `get`, `describe` now should be able to use `WatchList` to list resources, rather than
+directly list resources page by page.
+
+At this stage, we will adopt a similar approach to `client-go`, using an environment variable to control whether
+to enable this experimental feature, and ensuring that it can revert to the previous behavior if it fails.
+
+<<[UNRESOLVED Will `WatchList` replace `List` completely]>>
+`WatchList` was first released in version 1.27 and is still in the alpha stage, with continuous improvements 
+over the subsequent releases. For a feature in such an early stage, it is too soon to say whether we will fully
+rely on it without other fallback mechanisms.
+
+In my view, due to the performance improvements, we should use `WatchList` by default instead of `List` in the
+long term. For the evolution strategy, we can follow the approach of `client-go` and address it in a later release.
```

Comment:
A point to capture: external API servers may serve collections but only support the legacy **list** verb to fetch these.

Reply:
I think this is roughly consistent with kube-apiserver, and the key point is how external API handle **watch** request and `sendInitialEvents` param. 

If an error is returned, the fallback logic should always work. But if external API treat it as an normal **watch** request and not send `k8s.io/initial-events-end=true` event, thing goes wrong. 

If external API uses `k8s.io/apiserver` lib, it can get **watchlist** support for free by upgrade the lib to the latest.